### DISCUSSION TITLE
[Feature] Conservative with changing public Python API's

### DISCUSSION THREAD

**Discussion by JCZuurmond** [dbt-labs/dbt-core#10255] [General]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

Hi, I would like dbt to be conservative with public Python API's, i.e. minimize adding, removing and changing public Python API's.

This is more a "way-of-working" or "software engineering standards" request, then an typical feature request. (I did not consider it to be a bug either, maybe it could be a discussion instead of a feature.)

As a [package maintainer](https://github.com/godatadriven/pytest-dbt-core), I feel the pain when dbt changes public Python API's. For example, [see code here that handles the changing API's](https://github.com/godatadriven/pytest-dbt-core/pull/58):

``` python
try:
    from dbt.version import semver
except ImportError:
    try:
        import dbt_common.semver as semver
    except ImportError:
        try:
            import dbt.common.semver as semver
        except ImportError:
            import dbt.semver as semver
```

or

``` python
from dbt.adapters.factory import register_adapter,
from dbt import version

if semver.VersionSpecifier("1", "7", "16") < version.get_installed_version():
    from dbt.mp_context import get_mp_context
    register_adapter(config, get_mp_context())
else:
    register_adapter(config)
```

### Describe alternatives you've considered

Handling differences in Python API between dbt versions with Python logic.

### Who will this benefit?

Those who build tooling using the dbt Python API.

### Are you interested in contributing this feature?

No

### Anything else?

- Could be part of #6706
- Related to #10176

**Comment by dbeatty10** [comment]
> This is more a "way-of-working" or "software engineering standards" request, then an typical feature request. (I did not consider it to be a bug either, maybe it could be a discussion instead of a feature.)

Definitely a great topic for a discussion, so I'm going to convert it.
